File: event-sourcing/typescript/src/core/repository.ts
/**
 * Repository pattern for loading and saving aggregates
 */

import { AggregateId } from '../types/common';
import { Aggregate } from './aggregate';
import { ConcurrencyConflictError, EventStoreError } from './errors';
import type { EventStoreClient } from '../client/event-store-client';

/** Repository interface for aggregate persistence */
export interface Repository<TAggregate extends Aggregate> {
  /** Load an aggregate by ID */
  load(aggregateId: AggregateId): Promise<TAggregate | null>;

  /** Save an aggregate */
  save(aggregate: TAggregate): Promise<void>;

  /** Check if an aggregate exists */
  exists(aggregateId: AggregateId): Promise<boolean>;
}

/** Repository implementation using the event store */
export class EventStoreRepository<TAggregate extends Aggregate> implements Repository<TAggregate> {
  constructor(
    private eventStoreClient: EventStoreClient,
    private aggregateFactory: () => TAggregate,
    private aggregateType: string
  ) {}

  /** Load an aggregate by ID */
  async load(aggregateId: AggregateId): Promise<TAggregate | null> {
    try {
      const streamName = this.getStreamName(aggregateId);

      // Check if stream exists
      const exists = await this.eventStoreClient.streamExists(streamName);
      if (!exists) {
        return null;
      }

      // Read events from the stream
      const events = await this.eventStoreClient.readEvents(streamName);

      if (events.length === 0) {
        return null;
      }

      // Create aggregate and load from events
      const aggregate = this.aggregateFactory();
      aggregate.rehydrate(events);

      return aggregate;
    } catch (error) {
      throw new EventStoreError(
        `Failed to load aggregate ${this.aggregateType}:${aggregateId}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /** Save an aggregate */
  async save(aggregate: TAggregate): Promise<void> {
    if (!aggregate.id) {
      throw new Error('Cannot save aggregate without an ID');
    }

    if (!aggregate.hasUncommittedEvents()) {
      return; // Nothing to save
    }

    try {
      const streamName = this.getStreamName(aggregate.id);
      const events = aggregate.getUncommittedEvents();

      // Calculate expected version (current version - number of new events)
      const expectedVersion = aggregate.version - events.length;

      // Append events to the stream
      await this.eventStoreClient.appendEvents(streamName, events, expectedVersion);

      // Mark events as committed
      aggregate.markEventsAsCommitted();
    } catch (error) {
      if (this.isConcurrencyError(error)) {
        const pendingEvents = aggregate.getUncommittedEvents().length;
        const expectedVersion = aggregate.version - pendingEvents;
        throw new ConcurrencyConflictError(expectedVersion, aggregate.version);
      }

      throw new EventStoreError(
        `Failed to save aggregate ${this.aggregateType}:${aggregate.id}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /** Check if an aggregate exists */
  async exists(aggregateId: AggregateId): Promise<boolean> {
    try {
      const streamName = this.getStreamName(aggregateId);
      return await this.eventStoreClient.streamExists(streamName);
    } catch (error) {
      throw new EventStoreError(
        `Failed to check existence of aggregate ${this.aggregateType}:${aggregateId}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  /** Get the stream name for an aggregate */
  private getStreamName(aggregateId: AggregateId): string {
    return `${this.aggregateType}-${aggregateId}`;
  }

  /** Check if an error is a concurrency error */
  private isConcurrencyError(error: unknown): boolean {
    if (!error) {
      return false;
    }

    if (error instanceof ConcurrencyConflictError) {
      return true;
    }

    if (error instanceof Error) {
      const message = error.message.toLowerCase();
      return (
        message.includes('concurrency') ||
        message.includes('expected version') ||
        message.includes('precondition')
      );
    }

    if (typeof error === 'object' && error !== null) {
      const code = (error as { code?: unknown }).code;
      if (typeof code === 'string') {
        return code.toUpperCase() === 'FAILED_PRECONDITION' || code.toUpperCase() === 'ABORTED';
      }
    }

    return false;
  }
}

/** Repository factory for creating repositories */
export class RepositoryFactory {
  constructor(private eventStoreClient: EventStoreClient) {}

  /** Create a repository for a specific aggregate type */
  createRepository<TAggregate extends Aggregate>(
    aggregateFactory: () => TAggregate,
    aggregateType: string
  ): Repository<TAggregate> {
    return new EventStoreRepository(this.eventStoreClient, aggregateFactory, aggregateType);
  }
}

/** Repository registry for managing multiple repositories */
type AnyAggregate = Aggregate;
type AnyRepository = Repository<AnyAggregate>;

export class RepositoryRegistry {
  private repositories = new Map<string, AnyRepository>();

  /** Register a repository */
  register<TAggregate extends Aggregate>(
    aggregateType: string,
    repository: Repository<TAggregate>
  ): void {
    this.repositories.set(aggregateType, repository as AnyRepository);
  }

  /** Get a repository by aggregate type */
  get<TAggregate extends Aggregate>(aggregateType: string): Repository<TAggregate> | null {
    const repository = this.repositories.get(aggregateType) as Repository<TAggregate> | undefined;
    return repository ?? null;
  }

  /** Get a repository by aggregate type (throws if not found) */
  getRequired<TAggregate extends Aggregate>(aggregateType: string): Repository<TAggregate> {
    const repository = this.get<TAggregate>(aggregateType);
    if (!repository) {
      throw new Error(`No repository registered for aggregate type: ${aggregateType}`);
    }
    return repository;
  }
}

File: event-sourcing/python/src/event_sourcing/core/repository.py
"""Repository pattern for loading and saving aggregates."""

import logging
from collections.abc import Callable
from typing import Any, Generic, Protocol, TypeVar

from event_sourcing.core.aggregate import BaseAggregate

logger = logging.getLogger(__name__)

TAggregate = TypeVar("TAggregate", bound=BaseAggregate[Any])


class Repository(Protocol, Generic[TAggregate]):
    """Repository interface for aggregate persistence."""

    async def load(self, aggregate_id: str) -> TAggregate | None:
        """
        Load an aggregate by ID.

        Args:
            aggregate_id: The aggregate ID

        Returns:
            The loaded aggregate or None if not found
        """
        ...

    async def save(self, aggregate: TAggregate) -> None:
        """
        Save an aggregate with optimistic concurrency control.

        Args:
            aggregate: The aggregate to save

        Raises:
            ConcurrencyConflictError: If version conflict detected
        """
        ...

    async def exists(self, aggregate_id: str) -> bool:
        """
        Check if an aggregate exists.

        Args:
            aggregate_id: The aggregate ID

        Returns:
            True if the aggregate exists
        """
        ...


class EventStoreRepository(Generic[TAggregate]):
    """
    Event store implementation of the repository pattern.

    Loads aggregates by replaying their event history and saves them
    by appending new events to the stream.
    """

    def __init__(
        self,
        event_store_client: Any,  # EventStoreClient
        aggregate_factory: Callable[[], TAggregate],
        aggregate_type: str,
    ) -> None:
        """
        Initialize the repository.

        Args:
            event_store_client: Event store client for persistence
            aggregate_factory: Factory function to create new aggregate instances
            aggregate_type: The aggregate type name (used in stream names)
        """
        self._client = event_store_client
        self._aggregate_factory = aggregate_factory
        self._aggregate_type = aggregate_type

    def _get_stream_name(self, aggregate_id: str) -> str:
        """Get stream name for an aggregate."""
        return f"{self._aggregate_type}-{aggregate_id}"

    async def load(self, aggregate_id: str) -> TAggregate | None:
        """
        Load an aggregate by replaying its event history.

        Args:
            aggregate_id: The aggregate ID

        Returns:
            The loaded aggregate or None if not found
        """
        stream_name = self._get_stream_name(aggregate_id)

        # Read events from the stream
        events = await self._client.read_events(stream_name)

        if not events:
            logger.debug(f"Aggregate not found: {self._aggregate_type}:{aggregate_id}")
            return None

        # Create new aggregate instance
        aggregate = self._aggregate_factory()

        # Rehydrate from events
        aggregate.rehydrate(events)

        logger.debug(
            f"Loaded aggregate {self._aggregate_type}:{aggregate_id} at version {aggregate.version}"
        )

        return aggregate

    async def save(self, aggregate: TAggregate) -> None:
        """
        Save an aggregate by appending uncommitted events.

        Args:
            aggregate: The aggregate to save

        Raises:
            ConcurrencyConflictError: If version conflict detected
            InvalidAggregateStateError: If aggregate has no ID
        """
        if aggregate.id is None:
            from event_sourcing.core.errors import InvalidAggregateStateError

            raise InvalidAggregateStateError(
                self._aggregate_type, "Cannot save aggregate without ID"
            )

        # Get uncommitted events
        uncommitted_events = aggregate.get_uncommitted_events()

        if not uncommitted_events:
            logger.debug(f"No uncommitted events for {self._aggregate_type}:{aggregate.id}")
            return

        stream_name = self._get_stream_name(aggregate.id)

        # Calculate expected version (current version - uncommitted events)
        expected_version = aggregate.version - len(uncommitted_events)

        # Append events with optimistic concurrency
        await self._client.append_events(
            stream_name=stream_name,
            events=uncommitted_events,
            expected_version=expected_version,
        )

        # Mark events as committed
        aggregate.mark_events_as_committed()

        logger.debug(
            f"Saved aggregate {self._aggregate_type}:{aggregate.id} "
            f"with {len(uncommitted_events)} event(s), new version: {aggregate.version}"
        )

    async def exists(self, aggregate_id: str) -> bool:
        """Check if an aggregate exists."""
        stream_name = self._get_stream_name(aggregate_id)
        result: bool = await self._client.stream_exists(stream_name)
        return result


class RepositoryFactory:
    """Factory for creating event store repositories."""

    def __init__(self, event_store_client: Any) -> None:  # EventStoreClient
        """
        Initialize the factory.

        Args:
            event_store_client: Event store client for all repositories
        """
        self._client = event_store_client

    def create_repository(
        self,
        aggregate_factory: type[TAggregate],
        aggregate_type: str | None = None,
    ) -> EventStoreRepository[TAggregate]:
        """
        Create a repository for a specific aggregate type.

        Args:
            aggregate_factory: Aggregate class (used as factory)
            aggregate_type: Optional aggregate type name (defaults to class name)

        Returns:
            EventStoreRepository for the aggregate type
        """
        # Use class name if aggregate_type not provided
        if aggregate_type is None:
            aggregate_type = aggregate_factory.__name__.replace("Aggregate", "")

        # Create factory function
        def factory() -> TAggregate:
            return aggregate_factory()

        return EventStoreRepository(
            event_store_client=self._client,
            aggregate_factory=factory,
            aggregate_type=aggregate_type,
        )
